src/lib.rs

```diff
@@ -1,19 +1,521 @@
+//!
+//! Allow localhost to access your orbit using their temporary session key: did:key:z6Mktud6LcDFb3heS7FFWoJhiCafmUPkCAgpvJLv5E6fgBJg#z6Mktud6LcDFb3heS7FFWoJhiCafmUPkCAgpvJLv5E6fgBJg
+//!
+//! URI: did:key:z6Mktud6LcDFb3heS7FFWoJhiCafmUPkCAgpvJLv5E6fgBJg#z6Mktud6LcDFb3heS7FFWoJhiCafmUPkCAgpvJLv5E6fgBJg
+//! Version: 1
+//! Chain ID: 1
+//! Nonce: PPrtjztx2lYqWbqNs
+//! Issued At: 2021-12-20T12:29:25.907Z
+//! Expiration Time: 2021-12-20T12:44:25.906Z
+//! Resources:
+//! - kepler://bafk2bzacecn2cdbtzho72x4c62fcxvcqj23padh47s5jyyrv42mtca3yrhlpa#put
+//! - kepler://bafk2bzacecn2cdbtzho72x4c62fcxvcqj23padh47s5jyyrv42mtca3yrhlpa#del
+//! - kepler://bafk2bzacecn2cdbtzho72x4c62fcxvcqj23padh47s5jyyrv42mtca3yrhlpa#get
+//! - kepler://bafk2bzacecn2cdbtzho72x4c62fcxvcqj23padh47s5jyyrv42mtca3yrhlpa#list"#;
+//!
+//! const SIGNATURE: [u8; 65] = hex!("20c0da863b3dbfbb2acc0fb3b9ec6daefa38f3f20c997c283c4818ebeca96878787f84fccc25c4087ccb31ebd782ae1d2f74be076a49c0a8604419e41507e9381c");
+//!
+//! # fn run() -> anyhow::Result<siwe::Message> {
+//! # Ok({
+//! // `SignInWithEthereum` is the main entry point for both building and validation.
+//! let verifier = SignInWithEthereum::default()
+//!
+//!     // Common options for both building and validating messages:
+//!     .domain("localhost")?
+//!     .uri("did:key:z6Mktud6LcDFb3heS7FFWoJhiCafmUPkCAgpvJLv5E6fgBJg#z6Mktud6LcDFb3heS7FFWoJhiCafmUPkCAgpvJLv5E6fgBJg")?
+//!     .chain_id(1)
+//!
+//!     // Indicate that we want to verify a message (not build one.)
+//!     .verifier(MESSAGE)?
+//!
+//!     // Optionally pick a custom time, or leave it as the current time.
+//!     .valid_at(OffsetDateTime::from_unix_timestamp(1640003425)?)
+//!
+//!     // Customize which fields to verify (or leave the defaults.)
+//!     .statement(Verify::Never);
+//!
+//! // It is important to check that this message's nonce has never been used.
+//! let nonce = check_for_nonce_reuse(verifier.nonce())?;
+//!
+//! verifier
+//!     .verify(&nonce, &SIGNATURE)?
+//! # })
+//! # }
+//! # run().unwrap();
+//! ```
 use core::{
     convert::Infallible,
     fmt::{self, Display, Formatter},
     str::FromStr,
 };
 use hex::FromHex;
-use http::uri::{Authority, InvalidUri};
+use http::{
+    uri::{Authority, InvalidUri},
+    Uri,
+};
 use iri_string::types::UriString;
+use std::convert::TryFrom;
 use thiserror::Error;
+use time::Duration;
 use time::OffsetDateTime;
 
 pub mod nonce;
 pub mod rfc3339;
 
 pub use rfc3339::TimeStamp;
 
+#[derive(Debug, Clone, Copy)]
+enum Time {
+    Offset(Duration),
+    Exact(OffsetDateTime),
+}
+
+impl Time {
+    fn resolve(self) -> Option<OffsetDateTime> {
+        match self {
+            Self::Offset(d) => OffsetDateTime::now_utc().checked_add(d),
+            Self::Exact(e) => Some(e),
+        }
+    }
+}
+
+impl Default for Time {
+    fn default() -> Self {
+        Self::Offset(Default::default())
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct SignInWithEthereum {
```

Comment:
Not sure about the point of it (we already having parsing/validation management)

Reply:
My idea here is that you can share the code to create `SignInWithEthereum` between the build and verify paths, so you know you aren't setting different options. Say:

```rust
fn shared() -> SignInWithEthereum {
SignInWithEthereum::default()
    .statement("Hello, World")
    .chain_id(1337)
}

fn build() -> Message {
    shared()
        .builder()
        /* omitted */
        .build()
}

fn verify(...) -> Message {
    shared()
        .verifier(msg)
        /* omitted */
        .verify()
}
```

That way you can be sure you're using the same chain id, statement, etc.

Reply:
IMO this isn't _that_ much more easy to use, but it does add more cognitive overhead. 
"What's a `SignInWithEthereum` struct?"
The answer isn't obvious so I don't think we should add it in the current state.